packages/squiggle-lang/src/rescript/GenericDist/GenericDist_GenericOperation.res

```diff
@@ -0,0 +1,156 @@
+    run(extra, fnName)
+  }
+
+  let toPointSet = r => {
+    switch reCall(~fnName=#fromDist(#toDist(#toPointSet), r), ()) {
+    | #Dist(#PointSet(p)) => Ok(p)
+    | #Error(r) => Error(r)
+    | _ => Error(ImpossiblePath)
+    }
+  }
+
+  let toSampleSet = r => {
+    switch reCall(~fnName=#fromDist(#toDist(#toSampleSet(sampleCount)), r), ()) {
+    | #Dist(#SampleSet(p)) => Ok(p)
+    | #Error(r) => Error(r)
+    | _ => Error(ImpossiblePath)
+    }
+  }
+
+  let scaleMultiply = (r, weight) =>
+    reCall(
+      ~fnName=#fromDist(#toDistCombination(#Pointwise, #Multiply, #Float(weight)), r),
+      (),
+    ) |> outputToDistResult
+
+  let pointwiseAdd = (r1, r2) =>
+    reCall(
+      ~fnName=#fromDist(#toDistCombination(#Pointwise, #Add, #Dist(r2)), r1),
+      (),
+    ) |> outputToDistResult
+
+  let fromDistFn = (subFn: GenericDist_Types.Operation.fromDist, dist: genericDist) =>
+    switch subFn {
+    | #toFloat(fnName) =>
+      GenericDist.operationToFloat(toPointSet, fnName, dist)
+      |> E.R.fmap(r => #Float(r))
+      |> fromResult
+    | #toString => dist |> GenericDist.toString |> (r => #String(r))
+    | #toDist(#consoleLog) => {
+        Js.log2("Console log requested: ", dist)
+        #Dist(dist)
+      }
+    | #toDist(#normalize) => dist |> GenericDist.normalize |> (r => #Dist(r))
+    | #toDist(#truncate(left, right)) =>
+      dist |> GenericDist.truncate(toPointSet, left, right) |> E.R.fmap(r => #Dist(r)) |> fromResult
+    | #toDist(#toPointSet) =>
+      dist
+      |> GenericDist.toPointSet(xyPointLength)
+      |> E.R.fmap(r => #Dist(#PointSet(r)))
+      |> fromResult
+    | #toDist(#toSampleSet(n)) =>
+      dist |> GenericDist.sampleN(n) |> E.R.fmap(r => #Dist(#SampleSet(r))) |> fromResult
+    | #toDistCombination(#Algebraic, _, #Float(_)) => #Error(NotYetImplemented)
+    | #toDistCombination(#Algebraic, operation, #Dist(dist2)) =>
+      dist
+      |> GenericDist.algebraicCombination(toPointSet, toSampleSet, operation, dist2)
+      |> E.R.fmap(r => #Dist(r))
+      |> fromResult
+    | #toDistCombination(#Pointwise, operation, #Dist(dist2)) =>
+      dist
+      |> GenericDist.pointwiseCombination(toPointSet, operation, dist2)
+      |> E.R.fmap(r => #Dist(r))
+      |> fromResult
+    | #toDistCombination(#Pointwise, operation, #Float(f)) =>
+      dist
+      |> GenericDist.pointwiseCombinationFloat(toPointSet, operation, f)
+      |> E.R.fmap(r => #Dist(r))
+      |> fromResult
+    }
+
+  switch fnName {
+  | #fromDist(subFn, dist) => fromDistFn(subFn, dist)
+  | #fromFloat(subFn, float) => reCall(~fnName=#fromDist(subFn, GenericDist.fromFloat(float)), ())
+  | #mixture(dists) =>
+    GenericDist.mixture(scaleMultiply, pointwiseAdd, dists) |> E.R.fmap(r => #Dist(r)) |> fromResult
+  }
+}
+
+let runFromDist = (extra, fnName, dist) => run(extra, #fromDist(fnName, dist))
+let runFromFloat = (extra, fnName, float) => run(extra, #fromFloat(fnName, float))
+
+let fmap = (
```

Comment:
The problem with GenericDist_Types this is a pretty weird ``fmap`` that relies on the core GenericOperation functionality. GenericOperation depends on GenericDist_Types.

fmap is really just a wrapper on GenericOperation's ``run`` function.

It's not exactly a fit for GenericDist.E.fmap, because the type isn't GenericDist, but instead, a more generic type (that allows for floats, errors, and strings). 

I think we could probably come up with a better way of organizing this, but am not sure what it is yet. (More suggestions welcome!) I'd agree it's messy.

Reply:
I was reading this again and I decided that I like `fmapResult`, since that's the structure you're threading into